client/client-lib/src/query.rs

```diff
@@ -0,0 +1,194 @@
+        Self {
+            current: CurrentConsensus::default(),
+        }
+    }
+}
+
+impl<R: Hash + Eq + Clone> QueryStrategy<R> for Retry404<R> {
+    fn process(&mut self, response: FedResponse<R>, max_evil: usize) -> QueryStep<R> {
+        match response {
+            (peer, Err(JsonRpcError::Call(RpcCallError::Custom(e)))) if e.code() == 404 => {
+                QueryStep::Request(HashSet::from([peer]))
+            }
+            response => self.current.process(response, max_evil),
+        }
+    }
+}
+
+/// Returns when f+1 responses are equal, retrying after every 2f+1 responses
+pub struct EventuallyConsistent<R> {
+    responses: HashSet<PeerId>,
+    current: CurrentConsensus<R>,
+}
+
+impl<R> Default for EventuallyConsistent<R> {
+    fn default() -> Self {
+        Self {
+            responses: HashSet::new(),
+            current: CurrentConsensus::default(),
+        }
+    }
+}
+
+impl<R: Hash + Eq + Clone> QueryStrategy<R> for EventuallyConsistent<R> {
+    fn process(&mut self, response: FedResponse<R>, max_evil: usize) -> QueryStep<R> {
+        let honest_threshold = max_evil * 2 + 1;
+        self.responses.insert(response.0);
+
+        match self.current.process(response, max_evil) {
+            QueryStep::Continue if self.responses.len() == honest_threshold => {
+                let result = QueryStep::Request(self.responses.clone());
+                self.responses.clear();
+                result
+            }
+            result => result,
+        }
+    }
+}
+
+/// Returns when f+1 responses are equal
+pub struct CurrentConsensus<R> {
+    pub results: HashMap<R, HashSet<PeerId>>,
+    pub errors: HashMap<PeerId, JsonRpcError>,
+}
+
+impl<R> Default for CurrentConsensus<R> {
+    fn default() -> Self {
+        Self {
+            results: HashMap::new(),
+            errors: HashMap::new(),
+        }
+    }
+}
+
+impl<R: Hash + Eq + Clone> QueryStrategy<R> for CurrentConsensus<R> {
+    fn process(&mut self, response: FedResponse<R>, max_evil: usize) -> QueryStep<R> {
+        match response {
+            (peer, Ok(result)) => {
+                let peers = self
+                    .results
+                    .entry(result)
+                    .or_insert_with(|| HashSet::from([peer]));
+                peers.insert(peer);
+            }
+            (peer, Err(error)) => {
+                self.errors.insert(peer, error);
+            }
+        }
+
+        for (result, peers) in self.results.iter() {
+            if peers.len() > max_evil {
+                return QueryStep::Finished(Ok(result.clone()));
+            }
+        }
+
+        if self.errors.len() > max_evil {
+            let (_, error) = self.errors.drain().next().expect("non-empty");
+            return QueryStep::Finished(Err(ApiError::RpcError(error)));
+        }
+
+        QueryStep::Continue
+    }
+}
+
+pub trait QueryStrategy<R> {
+    fn process(&mut self, response: FedResponse<R>, max_evil: usize) -> QueryStep<R>;
```

Comment:
IMO `max_evil` should be provided once in the constructor. Some strategies don't need it anyway.